analytic_engine/src/instance/wal_replayer.rs

```diff
@@ -0,0 +1,596 @@
+            let serial_exec = table_data.serial_exec.lock().await;
+            let serial_exec_ctx = SerialExecContext {
+                table_data: table_data.clone(),
+                serial_exec,
+            };
+            serial_exec_ctxs.insert(table_data.id, serial_exec_ctx);
+        }
+
+        // Split and replay logs.
+        loop {
+            let decoder = WalDecoder::default();
+            log_entry_buf = log_iter
+                .next_log_entries(decoder, log_entry_buf)
+                .await
+                .box_err()
+                .context(ReplayWalWithCause { msg: None })?;
+
+            if log_entry_buf.is_empty() {
+                break;
+            }
+
+            Self::replay_single_batch(
+                context,
+                &log_entry_buf,
+                &mut serial_exec_ctxs,
+                table_results,
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+
+    async fn replay_single_batch(
+        context: &ReplayContext,
+        log_batch: &VecDeque<LogEntry<ReadPayload>>,
+        serial_exec_ctxs: &mut HashMap<TableId, SerialExecContext<'_>>,
+        table_results: &mut HashMap<TableId, Result<()>>,
+    ) -> Result<()> {
+        let mut table_batches = Vec::with_capacity(serial_exec_ctxs.len());
+        Self::split_log_batch_by_table(log_batch, &mut table_batches);
+        for table_batch in table_batches {
+            // Some tables may have failed in previous replay, ignore them.
+            let table_result = table_results.get(&table_batch.table_id);
+            if let Some(Err(_)) = table_result {
+                return Ok(());
+            }
+
+            // Replay all log entries of current table.
+            // Some tables may have been moved to other shards or dropped, ignore such logs.
+            if let Some(ctx) = serial_exec_ctxs.get_mut(&table_batch.table_id) {
+                let result = replay_table_log_entries(
+                    &context.flusher,
+                    context.max_retry_flush_limit,
+                    &mut ctx.serial_exec,
+                    &ctx.table_data,
+                    log_batch.range(table_batch.start_log_idx..table_batch.end_log_idx),
+                )
+                .await;
+                table_results.insert(table_batch.table_id, result);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn split_log_batch_by_table<P>(
+        log_batch: &VecDeque<LogEntry<P>>,
+        table_batches: &mut Vec<TableBatch>,
+    ) {
+        table_batches.clear();
+
+        if log_batch.is_empty() {
+            return;
+        }
+
+        // Split log batch by table id, for example:
+        // input batch:
+        //  |1|1|2|2|2|3|3|3|3|
+        //
+        // output batches:
+        //  |1|1|, |2|2|2|, |3|3|3|3|
+        let mut start_log_idx = 0usize;
+        let mut curr_log_idx = 0usize;
+        let mut start_table_id = log_batch.get(start_log_idx).unwrap().table_id;
+        loop {
```

Comment:
Could you try using [group by](https://doc.rust-lang.org/std/primitive.slice.html#method.group_by) to simplify this loop?

Reply:
No group by method in `VecDeque`...